feat!: export `ContractClient` et al. in `contract`; rename `SorobanRpc` to `rpc`

chadoh opened the pull request:
### Breaking Changes

- `ContractClient` functionality previously added in [v11.3.0](https://github.com/stellar/js-stellar-sdk/releases/tag/v11.3.0) was exported in a non-standard way. You can now import it as any other stellar-sdk module.

  ```diff
  - import { ContractClient } from '@stellar/stellar-sdk/lib/contract_client'
  + import { contract } from '@stellar/stellar-sdk'
  + const { Client } = contract
  ```

  Note that this top-level `contract` export is a container for ContractClient and related functionality. The ContractClient class is now available at `contract.Client`, as shown. Further note that there is a capitalized `Contract` export as well, which comes [from stellar-base](https://github.com/stellar/js-stellar-base/blob/b96281b9b3f94af23a913f93bdb62477f5434ccc/src/contract.js#L6-L19). You can remember which is which because capital-C `Contract` is a class, whereas lowercase-c `contract` is a container/module with a bunch of classes, functions, and types.

  Additionally, this is available from the `/contract` entrypoint, if your version of Node [and TypeScript](https://stackoverflow.com/a/70020984/249801) support [the `exports` declaration](https://nodejs.org/api/packages.html#exports). Finally, some of its exports have been renamed.

  ```diff
   import {
  -  ContractClient,
  +  Client,
     AssembledTransaction,
  -  ContractClientOptions,
  +  ClientOptions,
     SentTransaction,
  -} from '@stellar/stellar-sdk/lib/contract_client'
  +} from '@stellar/stellar-sdk/contract'
  ```


- The `ContractSpec` class is now nested under the `contract` module, and has been renamed to `Spec`.

  ```diff
  -import { ContractSpec } from '@stellar/stellar-sdk'
  +import { contract } from '@stellar/stellar-sdk'
  +const { Spec } = contract
  ```

  Alternatively, you can import this from the `contract` entrypoint, if your version of Node [and TypeScript](https://stackoverflow.com/a/70020984/249801) support [the `exports` declaration](https://nodejs.org/api/packages.html#exports).

  ```diff
  -import { ContractSpec } from '@stellar/stellar-sdk'
  +import { Spec } from '@stellar/stellar-sdk/contract'
  ```

- Previously, `AssembledTransaction.signAndSend()` would return a `SentTransaction` even if the transaction never finalized. That is, if it successfully sent the transaction to the network, but the transaction was still `status: 'PENDING'`, then it would `console.error` an error message, but return the indeterminate transaction anyhow.

  It now throws a `SentTransaction.Errors.TransactionStillPending` error with that error message instead.

### Deprecated

- `SorobanRpc` module is now also exported as `rpc`. You can import it with either name for now, but `SorobanRpc` will be removed in a future release.

  ```diff
   import { SorobanRpc } from '@stellar/stellar-sdk'
  +// OR
  +import { rpc } from '@stellar/stellar-sdk'
  ```

  You can also now import it at the `/rpc` entrypoint, if your version of Node [and TypeScript](https://stackoverflow.com/a/70020984/249801) support [the `exports` declaration](https://nodejs.org/api/packages.html#exports).

  ```diff
  -import { SorobanRpc } from '@stellar/stellar-sdk'
  -const { Api } = SorobanRpc
  +import { Api } from '@stellar/stellar-sdk/rpc'
  ```
  
### Other

* This PR started as an attempt to appease eslint for the `contract` directory (previously the `contract_client` directory). While this [turned out to be a bit premature](https://github.com/stellar/js-stellar-sdk/pull/962#issuecomment-2105016523) (given that much other code in this repo doesn't pass our eslint rules), it's still nice cleanup. That's why you'll see `.eslintrc.js` changes and related code updates. Some notes on all this:
  * We had `.eslintrc.js`, `src/.eslintrc.js`, and `src/soroban/.eslintrc.js`. The second two were very similar. I made them identical and removed the more deeply nested one. 
  * We have all TypeScript in `src`, but were only using Airbnb's non-TypeScript recommended settings. They recommend using both, for TypeScript projects. That's what I've done.
  * The [valid-jsdoc](https://eslint.org/docs/latest/rules/valid-jsdoc) rule was deprecated. They now recommend using the `eslint-plugin-jsdoc`. I've done that, and extended their recommended rules. I've overridden some rules. For example, I [researched and found](https://github.com/stellar/js-stellar-sdk/pull/962#discussion_r1596950321) that documenting function params with `@param` doesn't actually show up in your in-editor typeahead/mouseover. So I turned off the rule that requires these. We should prefer TSDoc for params instead. I've added these throughout.
* I attempted to add some JSDoc comments to the `export * as blah from 'whatever'` lines in `src/index.ts`. These don't actually show up when you import these objects. We will need to [research](https://jsdoc.app/howto-es2015-modules) this more when we do a larger module reorganization later. For now, they're useful notes.

chadoh commented on src/.eslintrc.js:
```diff
@@ -21,14 +24,7 @@ module.exports = {
     "no-debugger": 1,
-    "no-unused-vars": 1,
     "arrow-body-style": 1,
-    "valid-jsdoc": [
```
The docs for this plugin say that it is deprecated, and to favor `eslint-plugin-jsdoc` https://eslint.org/docs/latest/rules/valid-jsdoc

I think that's why the `requireReturnDescription` rule was not working correctly

chadoh commented on src/contract_client/basic_node_signer.ts:
```diff
@@ -1,21 +1,33 @@
  * applications. Feel free to use this as a starting point for your own
  * Wallet/TransactionSigner implementation.
+ * @param keypair {@link Keypair} to use to sign the transaction or auth entry
+ * @param networkPassphrase passphrase of network to sign for
  */
-export const basicNodeSigner = (keypair: Keypair, networkPassphrase: string) => ({
+export const basicNodeSigner = (
+  /**
+   *{@link Keypair} to use to sign the transaction or auth entry
+   */
+  keypair: Keypair,
+  /**
+   * passphrase of network to sign for
+   */
+  networkPassphrase: string,
```
I'm using both jsdoc syntax and TS syntax to document these params because I'm not sure which we should favor. I've more frequently used the TS version throughout, but our linter wants me to use jsdoc instead, and it doesn't care about the TS syntax.

Shaptic replied:
we use jsdoc almost everywhere (because reasons? iunno) and `yarn docs` might dgaf about ts syntax, so that's (unfortunately) preferred :pray:

chadoh replied:
See [research](https://github.com/stellar/js-stellar-sdk/pull/962#discussion_r1596950321) – I think we should prefer TSDoc `/** inline param docs */` from now on!

Shaptic commented on src/contract_client/basic_node_signer.ts:
```diff
@@ -1,21 +1,33 @@
-  }
-})
+  // eslint-disable-next-line require-await
+  signAuthEntry: async (entryXdr: string): Promise<string> =>
```
wait this is a good callout, if only for my own clarification - why is it marked `async`? isn't that implied by `Promise`?

chadoh replied:
it's that whole special compile-time-vs-runtime thing! The `Promise` is _TypeScript_, and gets stripped out before any JavaScript execution (or transpilation). The `async` is JavaScript. If you have an `async` function and forget to wrap the return type in a `Promise`, TypeScript will yell at you. You gotta keep 'em in sync!

Shaptic commented on src/contract_client/types.ts:
```diff
@@ -1,11 +1,8 @@
-  Transaction,
-  xdr,
-} from "..";
+/* eslint-disable @typescript-eslint/naming-convention */
```
if you're going to disable, a comment why is super helpful for review & posterity :pray:

chadoh replied:
I'll add a code comment, too, but the linter wants the type-naming convention to follow PascalCase. It doesn't like that we have a bunch of types like `i32` and `XDR_BASE64` in here. Renaming them all would be a big breaking change.

Shaptic commented on src/contract_client/utils.ts:
```diff
@@ -6,12 +8,32 @@ export const DEFAULT_TIMEOUT = 5 * 60;
+ * @param verbose Whether to log extra info
  */
 export async function withExponentialBackoff<T>(
+  /**
+   * Function to call repeatedly
+   */
```
echoing above: I can't imagine repeating yourself is required here :skull: but if it is, this format for one-liners is less noisy
```suggestion
  /** Function to call repeatedly */
```

Shaptic commented on .eslintrc.js:
```diff
@@ -2,9 +2,17 @@ module.exports = {
     es6: true,
   },
-  extends: ["airbnb-base", "prettier"],
+  extends: [
+    "airbnb-base",
+    "airbnb-typescript/base",
+    "prettier",
+    "plugin:jsdoc/recommended",
```
it'd be absolutely divine if you explained these changes for my own edification

chadoh replied:
Just here or also in the commit message?

- Extended `airbnb-typescript/base` to get it to stop yelling at me about importing files without file extension. Saw this recommended as the fix [on StackOverflow](https://stackoverflow.com/a/67610259/249801). And it makes sense to me that if we are extending Airbnb's lint rules and using TypeScript, we probably want their TypeScript-specific lint rules, too.

chadoh replied:
Also added these notes to the commit message

Shaptic reviewed (commented):
absolutely Herculean :clap: coupla requests :arrow_down:

leighmcculloch commented on src/contract_client/sent_transaction.ts:
```diff
@@ -118,7 +133,7 @@ export class SentTransaction<T> {
       this.getTransactionResponse.status ===
       SorobanRpc.Api.GetTransactionStatus.NOT_FOUND
     ) {
-      console.error(
+      throw new SentTransaction.Errors.TransactionStillPending(
```
Does this make the change a breaking change? If so we might need to wait for the next release?

leighmcculloch replied:
Or not actually because the release is an rc.

Shaptic replied:
yeah an rc makes it kosher but also `ContractClient` isn't published by the SDK so it's also kosher there
it'd be a breaking change for the cli/ts bindings but that is still safe because of the rc

chadoh replied:
Yeah, the linter didn't like the `console` statement, which made me stop and think—do we actually want it to just return, with just a `console.error` statement, if things are not actually in an expected state? If we're in an unexpected state, then we probably want to throw an error.

Shaptic replied:
Should this be noted in the changelog too?

chadoh replied:
It has now been noted in the changelog

leighmcculloch reviewed (commented):
Defer to @Shaptic, just one question from me.

chadoh commented on src/contract_client/basic_node_signer.ts:
```diff
@@ -1,21 +1,33 @@
+  keypair: Keypair,
+  /**
+   * passphrase of network to sign for
+   */
```
I'm guessing the jsdoc is shown in typeahead. I'll verify and get rid of the inline tsdoc-style comments.

chadoh replied:
ok, pretty disappointed rn, ngl

In order to tell where the docs are coming from, JSDoc or TSDoc, I changed the docs on `signAndSend` to this:

```ts
  /**
   * Sign the transaction with the `wallet`, included previously. If you did
   * not previously include one, you need to include one now that at least
   * includes the `signTransaction` method. After signing, this method will
   * send the transaction to the network and return a `SentTransaction` that
   * keeps track of all the attempts to fetch the transaction.
   * @param root0 JSDoc: options, lol
   * @param root0.force JSDoc: If `true`, sign and send the transaction even if it is a read call.
   * @param root0.signTransaction JSDOC: You must provide this here if you did not provide one before
   */
  signAndSend = async ({
    force = false,
    signTransaction = this.options.signTransaction,
  }: {
    /**
     * TSDoc: If `true`, sign and send the transaction even if it is a read call.
     */
    force?: boolean;
    /**
     * TSDoc: You must provide this here if you did not provide one before
     */
    signTransaction?: ContractClientOptions["signTransaction"];
  } = {}): Promise<SentTransaction<T>> => {
```

Note that this JSDoc `@param` notation, with stuff like `root0.force`, [is what they recommend doing](https://jsdoc.app/tags-param). Keep that in mind as you read on, so you can shake your head and chuckle appropriately.

Then I tried out typeahead / inline docs in Visual Studio Code, vim, and [Zed](https://zed.dev/). I'm loving the promise of Zed, but it's pretty new, and its typeahead experience was the worst. So let's not worry about it; barely anyone uses it anyhow.

Here's Vim's typeahead, using [Conquer of Completion with TypeScript](https://github.com/neoclide/coc-tsserver) (at the end you'll see me open the "hover" menu using a keyboard shortcut):

https://github.com/stellar/js-stellar-sdk/assets/221614/00fd2d6b-987d-4b0d-a00a-b61a80bd6ecf

And here's Visual Studio Code's:

https://github.com/stellar/js-stellar-sdk/assets/221614/e6059c8c-5437-421a-870f-0582bb602a87

You can see that **the JSDoc for the destructured params DOES NOT SHOW UP** in either case.

I think we should turn off the JSDoc param documentation, and prefer TSDoc style.

chadoh commented:
Didn't realize just how unhappy ESLint is with basically all the code in this repo 🤔

I saw that `src/.eslintrc.js` and `src/soroban/.eslintrc.js` were basically duplicates of each other. I tweaked them to match (removed the `no-unused-vars` setting, to make that an error case, which seems like what we actually want? One had been set to warn and the other had disabled it.), then removed `src/soroban/.eslintrc.js`.

I also fixed some config issues that I introduced earlier, so you can lint the whole repo again. And find that we have:

```
2081 problems (1063 errors, 1018 warnings)
```

¯\\\_(ツ)_/¯ 

Good enough for now, I guess??

Shaptic commented on src/contract_spec.ts:
```diff
@@ -1,13 +1,13 @@
   scValToBigInt,
-} from ".";
-import { Ok } from "./rust_types";
+} from "@stellar/stellar-base"
```
I think `from "."` correct here

Shaptic replied:
we should be very very careful to only reference base in one place, otherwise I'd be worried about the whole `instanceof` thing cropping up again

chadoh replied:
If you import `from "."` here, you get a "Dependency cycle detected (eslint: [import/no-cycle](https://github.com/import-js/eslint-plugin-import/blob/v2.29.1/docs/rules/no-cycle.md))" warning. Cyclic dependencies were causing all sorts of pain when I first exported `ContractClient`, though I can change this particular instance back and everything still builds. However, I noticed that other files throughout, [such as `src/soroban/api.ts`](https://github.com/AhaLabs/js-stellar-sdk/blob/2ee6126c72b36a458787df78b60913c579d42b3c/src/soroban/api.ts#L1), are importing from `@stellar/stellar-base`, which is why I started following suit.

Shaptic commented on src/soroban/contract_client/index.ts:
```diff
@@ -0,0 +1,7 @@
+export * from "./assembled_transaction";
+export * from "./basic_node_signer";
+export * from "./client";
+export * from "./rust_result";
+export * from "./sent_transaction";
+export * from "./types";
+export * from "./utils";
```
are we SURE all of these should be exported (e.g. `utils` feels excessive)? at least at the package level. remember that it means we have to commit to **every** **single** name, parameter type/count, method, etc.

chadoh replied:
Eh... I could honestly go either way. `utils` exports:

- `DEFAULT_TIMEOUT`
- `withExponentialBackoff`
- `contractErrorPattern`
- `implementsToString`

I could see all of these being useful, and they are all namespaced under `ContractClient`, so maybe it's ok?

But if we want to keep any in particular we could move the definition elsewhere and only export that.

chadoh replied:
But yeah, if we want to keep the implementation small(er), we can remove `utils`. I think we want to export the rest of it, though! It's all used by TS Bindings.

chadoh replied:
Removed `utils`, kept the rest.

Shaptic commented on src/soroban/index.ts:
```diff
@@ -2,12 +2,13 @@
+export { default as AxiosClient } from "./axios";
+export { parseRawSimulation, parseRawEvents } from "./parsers";
+export * from "./transaction";
+export * as ContractClient from "./contract_client";
```
noted this on Slack, but imo this shouldn't be in SorobanRpc. Coupla options, imo:

* rename `src/soroban/` to `src/soroban/rpc/` and leave it as-is, then move these to the higher level `src/soroban/`
  - you'd still have `StellarSdk.SorobanRpc` but there'd also be `StellarSdk.Soroban` (or, ideally, a better name, e.g. `ContractHelpers` or `Binding` or idk honestly) that contained all or a subset of `src/soroban/contract_client/`
* do this same thing but manually by `delete`ing and injecting stuff into the global export in `src/index.ts`

chadoh replied:
I don't have strong opinions on this. I definitely found it strange to lump this under _RPC_. I thought maybe we already had more Soroban stuff than just RPC stuff lumped into this namespace, so I figured maybe we didn't care. But I guess that's not true.

Could do something like:

- `StellarSdk.Soroban.Rpc`
- `StellarSdk.Soroban.Contract`
- `StellarSdk.Soroban.Contract.Client`
- `StellarSdk.Soroban.Contract.AssembledTransaction`
- etc

Not sure.

Whatever we pick, I want to make sure that we can import the types. Don't want people to need to do what Tyler needed to do with this PR today, using `SorobanRpc.ContractClient.ContractClientOptions` and similar everywhere, because importing those types from `SorobanRpc.ContractClient` didn't work.

chadoh replied:
We could also try using [the `exports` declaration](https://hackmd.io/@dzearing/ry1jTjCKd) to make importing these deeply-nested objects easier.

```ts
import { Client } from 'stellar-sdk/Contract'
import { Api } from 'stellar-sdk/Rpc'
```

Still not sure what the best API is here.

chadoh replied:
Note that in my examples above I dropped `Soroban` from the path. We probably want to do that eventually. Maybe now's the time? Seems like we really just have RPC stuff and ContractClient stuff? I think for the sake of imports it makes sense to import a _Client_ from a path called _Contract_, but I could also see doing `import { Client } from 'stellar-sdk/ContractClient'`

Shaptic commented on CHANGELOG.md:
````diff
@@ -6,6 +6,64 @@ A breaking change will get clearly marked in this log.
+
+  ```diff
+  - import { ContractClient } from '@stellar/stellar-sdk/lib/contract_client'
+  - import { ContractClient } from '@stellar/stellar-sdk'
````
typo: 
```suggestion
  + import { ContractClient } from '@stellar/stellar-sdk'
```

Shaptic commented on CHANGELOG.md:
```diff
@@ -6,6 +6,64 @@ A breaking change will get clearly marked in this log.
+
+### Added
+
+- `SorobanRpc` module is now also exported as `Rpc`. You can import it with either name.
```
Nice! I'm for this; can you also note that `SorobanRpc` is now deprecated and will be removed in the future?

chadoh replied:
Are we ready to officially deprecate it? Does that require anything more than mentioning it in the changelog?

Shaptic replied:
Just the latter, it'll be informal

chadoh replied:
Ok, changelog mentions that it's deprecated.

Shaptic commented on package.json:
```diff
@@ -23,6 +23,21 @@
+      "types": "./lib/rpc/index.d.ts",
+      "default": "./lib/rpc/index.js"
+    }
+  },
```
:fire:

Shaptic reviewed (commented):
A few small notes but otherwise this is :fire: :fire: :fire: great work! :clap:

chadoh commented:
I've confirmed that the new entrypoints are usable from the TS Bindings code in the CLI:

- https://github.com/stellar/soroban-cli/pull/1326

I think we need to make sure we like the new module organization and then we're good to go. The current reorganization is explained in the changelog entry. The gist:

- Add new `ContractClient` exported module, along with `@stellar/stellar-sdk/ContractClient` entrypoint
- Move `ContractSpec` to `ContractClient.Spec`, which can be imported as
   ```ts
   import { Spec } from '@stellar/stellar-sdk/ContractClient'
   ```
- Keep `SorobanRpc` as-is, though it is also now also exported as just `Rpc`, and is also available from its own entrypoint:
   ```ts
   import { Api } from '@stellar/stellar-sdk/Rpc'
   ```

Happy to roll this third thing back for now, since it's sort of irrelevant to the current change. I just wanted to see if it was possible, and it's a non-breaking change that could help us think through how to do a larger module reorganization after the next release is out.

Shaptic commented on test/unit/server/soroban/constructor_test.js:
```diff
@@ -1,4 +1,4 @@
-const { Server, AxiosClient } = StellarSdk.SorobanRpc;
+const { Server, AxiosClient } = StellarSdk.Rpc;
```
lets keep one of these as SorobanRpc so we ensure there's no backwards breakage :+1:
